### PR TITLE
Encode point estimate output type IDs with null. 

### LINKED ISSUE #109: Null point estimate output type IDs in configs instead of `NA`s
@zkamvar posted the following note on #103 

Something that was brought up in response to https://github.com/reichlab/variant-nowcast-hub/pull/117#issuecomment-2423170370 is that the `"NA"` is a bit confusing because it sure looks like a character, but when we expand the grid the `output_type_id` columns become `NA` (which is an intentional move by Ooms described in [section 2.1.1 of the JSONlite package paper](https://cran.r-project.org/web/packages/jsonlite/vignettes/json-mapping.pdf))

Now that we are using `is_required` for point estimate types, **we might be able to take this opportunity to set the `required` property to a single element `null` array.** This will have exactly the same result as the `"NA"` array, but with the following advantages:

1. inter-language compatibility can be achieved since `null` is a concept that even JSON can understand
2. we can clearly communicate that this should be a missing value in the data as opposed to a character string. 

This is what I think it would look like in the schema: 

```
"required": {
    "description": "Not relevant for point estimate output types. Must be a single array of null"
    "type": "array"
    "items": {
        "const": null,
        "maxItems": 1
    }
}
```

### Demo

Here's a demo that shows that `["NA"]` and `[null]` are equivalent by modifying a tasks.json file and reading them in with jsonlite

``` r
hub_con <- hubData::connect_hub(
  system.file("testhubs/simple", package = "hubUtils")
)
hp <- attributes(hub_con)$hub_path
# Get the tasks file which has a `mean` output type
tasks <- fs::path(hp, "hub-config", "tasks.json")

# copy the file to a new temp file
tmp <- withr::local_tempfile()
fs::file_copy(tasks, tmp, overwrite = TRUE)
# The two files are identical
unname(tools::md5sum(tmp) == tools::md5sum(tasks))
#> [1] TRUE

cfg <- readLines(tmp)
writeLines(cfg[231:240]) # optional is NA
#>                 "output_type": {
#>                     "mean": {
#>                         "output_type_id": {
#>                             "required": null,
#>                             "optional": ["NA"]
#>                         },
#>                         "value": {
#>                             "type": "integer",
#>                             "minimum": 0
#>                         }

# Change '["NA"]' to '[null]'
nullcfg <- sub('["NA"]', '[null]', cfg, fixed = TRUE)
writeLines(nullcfg[231:240]) # optional is now null
#>                 "output_type": {
#>                     "mean": {
#>                         "output_type_id": {
#>                             "required": null,
#>                             "optional": [null]
#>                         },
#>                         "value": {
#>                             "type": "integer",
#>                             "minimum": 0
#>                         }
writeLines(nullcfg, con = tmp) # changing to null

# The two files are no longer identical
unname(tools::md5sum(tmp) == tools::md5sum(tasks))
#> [1] FALSE
waldo::compare(cfg, nullcfg)
#> old[81:87] vs new[81:87]
#>   "                    \"mean\": {"
#>   "                        \"output_type_id\": {"
#>   "                            \"required\": null,"
#> - "                            \"optional\": [\"NA\"]"
#> + "                            \"optional\": [null]"
#>   "                        },"
#>   "                        \"value\": {"
#>   "                            \"type\": \"integer\","
#> 
#> old[232:238] vs new[232:238]
#>   "                    \"mean\": {"
#>   "                        \"output_type_id\": {"
#>   "                            \"required\": null,"
#> - "                            \"optional\": [\"NA\"]"
#> + "                            \"optional\": [null]"
#>   "                        },"
#>   "                        \"value\": {"
#>   "                            \"type\": \"integer\","


# The resulting R object after reading in JSON are identical
identical(
  jsonlite::fromJSON(tasks, simplifyDataFrame = FALSE), 
  jsonlite::fromJSON(tmp, simplifyDataFrame = FALSE)
)
#> [1] TRUE
```

<sup>Created on 2024-10-18 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

_Originally posted by @zkamvar in https://github.com/hubverse-org/schemas/issues/103#issuecomment-2423356765_

### PR DESCRIPTION
**Pull request by annakrystalli**
Resolves #109

### CODE REVIEW

**Comment by annakrystalli** [comment]
\diff

File: v4.0.0/tasks-schema.json [hunk 1]
```diff
@@ -652,23 +652,17 @@
                                             "description": "Object defining the mean of the predictive distribution output type.",
                                             "properties": {
                                                 "output_type_id": {
-                                                    "description": "output_type_id is not meaningful for a point estimate output_type.",
+                                                    "description": "output_type_id is not meaningful for a point estimate output_type. Must be null",
```
Comment (elray1) [inline v4.0.0/tasks-schema.json:655]:
```suggestion
                                                    "description": "output_type_id is not meaningful for a point estimate output_type. Must have a single property called 'required' with the value null.",
```

File: v4.0.0/tasks-schema.json [hunk 2]
```diff
@@ -736,23 +730,17 @@
                                             "description": "Object defining the median of the predictive distribution output type",
                                             "properties": {
                                                 "output_type_id": {
-                                                    "description": "output_type_id is not meaningful for a point estimate output_type.",
+                                                    "description": "output_type_id is not meaningful for a point estimate output_type. Must be null",
```
Comment (elray1) [inline v4.0.0/tasks-schema.json:733]:
```suggestion
                                                    "description": "output_type_id is not meaningful for a point estimate output_type. Must have a single property named 'required' with the value null.",
```

**Review by elray1 — APPROVED** [review]
I made a couple of minor wording suggestions, but leave it up to you what you think is clearer.  Approving.

**Review by zkamvar — APPROVED** [review]
I think setting `required: null` for the point estimates makes the most sense.

I also wanted to also provide context for what we can expect for adapting
`hubValidations` to work with this new format, becuase setting
`required: null` will make point estimates invisible to
`expand_model_out_grid` as demonstrated below. 

Setup: creating three configurations:

- `v3` the v3 config where mean has `required: ["NA"]`
- `array_null` the v3 config, but the mean has `required: [null]`
- `solo_null` the v3 config, but the mean has `required: null`

``` r
hub_con <- hubData::connect_hub(
  system.file("testhubs/simple", package = "hubUtils")
)
hp <- attributes(hub_con)$hub_path
# Get the tasks file which has a `mean` output type
tasks <- fs::path(hp, "hub-config", "tasks.json")
# copy the file to a new temp file
cfg <- readLines(tasks)
# change `required:` to a bare `null`
nullcfg <- sub('["NA"]', 'null', cfg, fixed = TRUE)
# change `required:` to a `[null]` array
null_array_cfg <- sub('["NA"]', '[null]', cfg, fixed = TRUE)
# convert to arrays
v3 <- jsonlite::fromJSON(cfg, simplifyDataFrame = FALSE)
array_null <- jsonlite::fromJSON(null_array_cfg, simplifyDataFrame = FALSE)
solo_null <- jsonlite::fromJSON(nullcfg, simplifyDataFrame = FALSE)
```

null array matches our current infrastructure, but bare null is different

``` r
identical(v3, array_null)
#> [1] TRUE
identical(v3, solo_null)
#> [1] FALSE
```

The implication for this is that `expand_model_out_grid()` no longer detects
point estimates. The omission happens in [expand_model_out_grid.R#214](https://github.com/hubverse-org/hubValidations/blob/5c8395297181e2a2cd1f73661155511278d48438/R/expand_model_out_grid.R#L197-L215)
with `purrr::compact()` throwing away `list(mean = NULL)`. I think we can
address this by adding a v4 catch at that point that protects point estimates.

``` r
library("hubValidations")
head(expand_model_out_grid(v3, "2022-10-01"))
#> # A tibble: 6 × 6
#>   origin_date target          horizon location output_type output_type_id
#>   <date>      <chr>             <int> <chr>    <chr>                <dbl>
#> 1 2022-10-01  wk inc flu hosp       1 US       mean                    NA
#> 2 2022-10-01  wk inc flu hosp       2 US       mean                    NA
#> 3 2022-10-01  wk inc flu hosp       3 US       mean                    NA
#> 4 2022-10-01  wk inc flu hosp       4 US       mean                    NA
#> 5 2022-10-01  wk inc flu hosp       1 01       mean                    NA
#> 6 2022-10-01  wk inc flu hosp       2 01       mean                    NA
head(expand_model_out_grid(array_null, "2022-10-01"))
#> # A tibble: 6 × 6
#>   origin_date target          horizon location output_type output_type_id
#>   <date>      <chr>             <int> <chr>    <chr>                <dbl>
#> 1 2022-10-01  wk inc flu hosp       1 US       mean                    NA
#> 2 2022-10-01  wk inc flu hosp       2 US       mean                    NA
#> 3 2022-10-01  wk inc flu hosp       3 US       mean                    NA
#> 4 2022-10-01  wk inc flu hosp       4 US       mean                    NA
#> 5 2022-10-01  wk inc flu hosp       1 01       mean                    NA
#> 6 2022-10-01  wk inc flu hosp       2 01       mean                    NA
head(expand_model_out_grid(solo_null, "2022-10-01"))
#> # A tibble: 6 × 6
#>   origin_date target          horizon location output_type output_type_id
#>   <date>      <chr>             <int> <chr>    <chr>                <dbl>
#> 1 2022-10-01  wk inc flu hosp       1 US       quantile              0.01
#> 2 2022-10-01  wk inc flu hosp       2 US       quantile              0.01
#> 3 2022-10-01  wk inc flu hosp       3 US       quantile              0.01
#> 4 2022-10-01  wk inc flu hosp       4 US       quantile              0.01
#> 5 2022-10-01  wk inc flu hosp       1 01       quantile              0.01
#> 6 2022-10-01  wk inc flu hosp       2 01       quantile              0.01
```

<sup>Created on 2024-10-25 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Comment by annakrystalli** [comment]
Thanks @zkamvar .

None of the proposed schema changes work yet because the functionality to support it hasn't been developed yet. 

The usual development process is to queue up the new version in `schema`  repo `br-newversion` branch and then build the new functionality against it. All we need to concern ourselves in this PR is the correctecness of the schema. The rest will be develop in the individual packages affected. I have plans for each that are separate for this.

**Pull request merged**